boutdata/collect.py

```diff
@@ -307,392 +239,823 @@ def getDataFile(i):
+        ranges = []
+    elif dimensions == ("t",):
+        ranges = [tind]
+    elif dimensions == ("x", "y"):
+        # Field2D
+        ranges = [xind, yind]
+    elif dimensions == ("x", "z"):
+        # FieldPerp
+        ranges = [xind, zind]
+    elif dimensions == ("t", "x", "y"):
+        # evolving Field2D
+        ranges = [tind, xind, yind]
+    elif dimensions == ("t", "x", "z"):
+        # evolving FieldPerp
+        ranges = [tind, xind, zind]
+    elif dimensions == ("x", "y", "z"):
+        # Field3D
+        ranges = [xind, yind, zind]
+    elif dimensions == ("t", "x", "y", "z"):
+        # evolving Field3D
+        ranges = [tind, xind, yind, zind]
+    else:
+        # Not a Field, so do not support slicing.
+        # For example, may be a string.
+        ranges = None
+
+    data = f.read(varname, ranges)
+    var_attributes = f.attributes(varname)
+    return BoutArray(data, attributes=var_attributes)
+
+
+def _read_scalar(f, varname, dimensions, var_attributes, tind):
+    """
+    Read a scalar variable from a single file
+
+    Parameters
+    ----------
+    f : DataFile
+        File to read from. This function does *not* close f.
+    varname : str
+        Name of variable to read
+    dimensions : tuple
+        Dimensions of the variable
+    var_attributes : dict
+        Attributes of the variable
+    tind : slice
+        Slice to apply to the t-dimension, if there is one
+    """
+    if 't' in dimensions:
+        if not dimensions[0] == 't':
+            # 't' should be the first dimension in the list if present
+            raise ValueError(
+                "{} has a 't' dimension, but it is not the first dimension "
+                "in dimensions={}".format(varname, dimensions)
+            )
+        data = f.read(varname, ranges = [tind] + (len(dimensions) - 1) * [None])
+    else:
+        # No time or space dimensions, so no slicing
+        data = f.read(varname)
+    return BoutArray(data, attributes=var_attributes)
+
+
+def _apply_step(data, dimensions, xstep, ystep):
+    """
+    Apply steps of xind and yind slices to an array
+
+    Parameters
+    ----------
+    data : np.Array
+        Data array to be sliced
+    dimensions : tuple
+        Dimensions of data
+    xstep : int or None
+        Step to apply in the x-direction
+    ystep : int or None
+        Step to apply in the y-direction
+    """
+    if xstep is not None or ystep is not None:
+        if dimensions == ("t", "x", "y", "z"):
+            data = data[:, :: xstep, :: ystep]
+        elif dimensions == ("x", "y", "z"):
+            data = data[:: xstep, :: ystep, :]
+        elif dimensions == ("t", "x", "y"):
+            data = data[:, :: xstep, :: ystep]
+        elif dimensions == ("t", "x", "z"):
+            data = data[:, :: xstep, :]
+        elif dimensions == ("x", "y"):
+            data = data[:: xstep, :: ystep]
+        elif dimensions == ("x", "z"):
+            data = data[:: xstep, :]
```

Comment:
Maybe something like this? It's a bit shorter, not sure if it's any more readable:

```python
# Array of slices the correct length, slice(None) means [:]
slices = [slice(None)] * len(dimensions.shape)

if xstep is not None:
    x_index = dimensions.index("x")
    slices[x_index] = slice(None, None, xstep)

if ystep is not None:
    y_index = dimensions.index("y")
    slices[y_index] = slice(None, None, ystep)

data = data[tuple(slices)]
```

`dimensions.index()` will raise `ValueError` if the element isn't found, so it might need to be wrapped in a `try` block